File: scripts/send_opencode_message.py
#!/usr/bin/env python3
"""Send a chat message into an active Opencode session (minimal example).

Usage:
  python send_opencode_message.py --text "Hello" [--base-url http://127.0.0.1:4096] [--session-id <id>]

Behavior:
 - If --session-id is provided the message is sent to that session.
 - Otherwise the script lists sessions and picks an "active" session or the first one.

Notes:
 - The Opencode SDK (opencode-ai) must be installed in the environment.
 - You can set OPENCODE_BASE_URL env var instead of --base-url.
 - This is best-effort: network errors are printed but do not raise uncaught.
"""

from __future__ import annotations

import argparse
import os
import sys
import traceback
import socket
import time
from urllib.parse import urlparse

import httpx
from typing import Any, cast

try:
    import opencode_ai
    from opencode_ai import Opencode
    from opencode_ai import APIConnectionError
except Exception:  # pragma: no cover - helpful error message for missing dependency
    print(
        "Missing dependency: install the official SDK 'opencode-ai' (pip install --pre opencode-ai)"
    )
    raise


def send_message(
    text: str,
    base_url: str | None = None,
    session_id: str | None = None,
    model_id: str | None = None,
    provider_id: str | None = None,
) -> int:
    """Send `text` into the given session or the first active session.

    Returns 0 on success, non-zero on failure.
    """
    client = Opencode(base_url=base_url) if base_url else Opencode()

    try:
        if session_id:
            target = session_id
        else:
            sessions = client.session.list()
            # sessions may be a list-like or an iterable; convert to list to inspect
            sessions = list(sessions)
            if not sessions:
                print("No sessions available on the Opencode instance.")
                return 2

            active = next(
                (s for s in sessions if getattr(s, "status", None) == "active"), None
            )
            selected = active or sessions[0]
            target = getattr(selected, "id", None) or getattr(
                selected, "session_id", None
            )
            if not target:
                print("Could not determine session id from session object:", selected)
                return 3

        # parts is the structured message payload used by the SDK
        parts = [{"type": "text", "text": text}]

        # If model/provider not provided, attempt to infer from the last message's info block
        if not model_id or not provider_id:
            try:
                msg = _get_last_message(client, target)
                inferred_model, inferred_provider = (
                    _extract_model_provider_from_message(msg)
                )
                if inferred_model and inferred_provider:
                    model_id = model_id or inferred_model
                    provider_id = provider_id or inferred_provider
                    print(
                        f"Inferred model_id={model_id} provider_id={provider_id} from last message"
                    )
                else:
                    print(
                        "Could not infer model_id/provider_id from recent messages.\n"
                        "Provide them via --model-id/--provider-id or set OPENCODE_MODEL_ID/OPENCODE_PROVIDER_ID env vars."
                    )
                    return 6
            except Exception:
                print(
                    "Failed to infer model/provider from last message; please provide --model-id and --provider-id"
                )
                return 6

        # Use session.prompt to create/send a new message in the session.
        # The SDK expects a `model` object for provider/model identifiers.
        model_obj = {"providerID": provider_id, "modelID": model_id}
        # Try preferred high-level SDK method first (session.prompt)
        try:
            if hasattr(client.session, "prompt"):
                client.session.prompt(
                    id=target, parts=cast(Any, parts), model=model_obj
                )
            else:
                raise AttributeError("session.prompt not available")
        except Exception as e:
            # Fallback: POST directly to the session prompt endpoint
            print(
                f"High-level send failed ({type(e).__name__}): {e}; trying low-level POST /session/{target}/prompt"
            )
            body = {
                "parts": parts,
                "model": {"providerID": provider_id, "modelID": model_id},
            }
            try:
                resp = client.post(f"/session/{target}/prompt", body=body, cast_to=dict)
                print("Low-level POST response type:", type(resp))
            except Exception as e2:
                print("Low-level POST failed:", type(e2).__name__, e2)
                raise
        print(f"Message sent to session {target}")
        return 0

    except APIConnectionError as e:
        print("Connection error while talking to Opencode API:", e)
        if getattr(e, "__cause__", None):
            print("Underlying cause:")
            traceback.print_exception(e.__cause__)
        return 4
    except Exception as e:
        print("Error sending message:", type(e).__name__, e)
        traceback.print_exc()
        return 5


def _diagnose_base_url(base_url: str) -> None:
    """Perform lightweight diagnostics for base_url: DNS, TCP connect and optional health endpoint.

    Prints findings to stdout to help debug connection issues.
    """
    print(f"Diagnostic: probing {base_url}")
    parsed = urlparse(base_url)
    host = parsed.hostname or parsed.path
    port = parsed.port or (443 if parsed.scheme == "https" else 80)
    print(f"Resolved host={host} port={port} scheme={parsed.scheme}")

    try:
        infos = socket.getaddrinfo(host, port, proto=socket.IPPROTO_TCP)
        addrs = sorted({str(ai[4][0]) for ai in infos})
        print("Address(es):", ", ".join(addrs))
    except Exception as e:
        print("getaddrinfo failed:", e)

    health_url = base_url.rstrip("/") + "/health"
    for attempt in range(1, 4):
        try:
            print(f"HTTP probe attempt {attempt} -> {health_url}")
            with httpx.Client(timeout=5.0) as c:
                r = c.get(health_url)
            print("HTTP probe status:", r.status_code)
            break
        except Exception as e:
            print(f"Probe attempt {attempt} failed: {e}")
            time.sleep(0.5)


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(
        description="Send a message into an Opencode session (minimal example)"
    )
    p.add_argument("--text", required=True, help="Text to send into the session")
    p.add_argument(
        "--base-url",
        default=os.environ.get("OPENCODE_BASE_URL", "http://127.0.0.1:4096"),
        help="Opencode base URL (default: http://127.0.0.1:4096)",
    )
    p.add_argument(
        "--debug",
        action="store_true",
        help="Enable lightweight diagnostics before sending",
    )
    p.add_argument(
        "--model-id",
        default=os.environ.get("OPENCODE_MODEL_ID"),
        help="Model id to use for session.chat (or set OPENCODE_MODEL_ID)",
    )
    p.add_argument(
        "--provider-id",
        default=os.environ.get("OPENCODE_PROVIDER_ID"),
        help="Provider id to use for session.chat (or set OPENCODE_PROVIDER_ID)",
    )
    p.add_argument(
        "--session-id",
        help="If provided, send to this session id instead of selecting an active session",
    )
    args = p.parse_args(argv)

    if args.debug and args.base_url:
        _diagnose_base_url(args.base_url)

    return send_message(
        text=args.text,
        base_url=args.base_url,
        session_id=args.session_id,
        model_id=args.model_id,
        provider_id=args.provider_id,
    )


def _get_last_message(client: Opencode, session_id: str):
    """Try to fetch recent messages for a session using a few possible call signatures.

    Returns the last message object or None.
    """
    # The SDK exposes session.messages(id=...) which returns a list of message items.
    try:
        res = client.session.messages(id=session_id)
        # res may already be a list
        try:
            messages = list(res)
        except Exception:
            messages = res
        if messages:
            return messages[-1]
    except TypeError:
        # signature mismatch; fall through
        pass
    except Exception:
        # network or other error
        return None

    # If above failed, try message resource if present
    msg_resource = getattr(client, "message", None)
    if not msg_resource:
        return None

    try:
        func = getattr(msg_resource, "list")
        messages = func(session_id=session_id)
        messages = list(messages)
        if messages:
            return messages[-1]
    except Exception:
        return None

    return None


def _extract_model_provider_from_message(msg) -> tuple[str | None, str | None]:
    """Extract model_id and provider_id from a message object `info` block.

    Returns (model_id, provider_id) or (None, None).
    """
    if not msg:
        return None, None

    # Try pydantic helpers first
    data = None
    try:
        data = msg.to_dict()
    except Exception:
        try:
            data = getattr(msg, "model_dump", lambda: None)()
        except Exception:
            data = getattr(msg, "__dict__", None)

    if not data or not isinstance(data, dict):
        return None, None

    # info blocks may live under several keys
    info = (
        data.get("info")
        or data.get("metadata")
        or data.get("meta")
        or data.get("extras")
    )
    if isinstance(info, dict):
        # accept multiple casing conventions used in the SDK responses
        model_id = (
            info.get("model_id")
            or info.get("modelId")
            or info.get("modelID")
            or info.get("model")
        )
        provider_id = (
            info.get("provider_id")
            or info.get("providerId")
            or info.get("providerID")
            or info.get("provider")
        )
        return model_id, provider_id

    # fallback: check top-level keys
    model_id = (
        data.get("model_id")
        or data.get("modelId")
        or data.get("modelID")
        or data.get("model")
    )
    provider_id = (
        data.get("provider_id")
        or data.get("providerId")
        or data.get("providerID")
        or data.get("provider")
    )
    return model_id, provider_id


if __name__ == "__main__":
    sys.exit(main())

File: agent_workspace/scripts/send_opencode_message.js
#!/usr/bin/env node
/*
 * Send a chat message into an active Opencode session (minimal JS port of
 * agent-workspace/scripts/send_opencode_message.py).
 *
 * Usage:
 *   node send_opencode_message.js --text "Hello" [--base-url http://127.0.0.1:4096] [--session-id <id>]
 */

import { createOpencodeClient } from "@opencode-ai/sdk";
import { URL } from "url";
import dns from "node:dns";
import net from "node:net";
import process from "node:process";

async function diagnoseBaseUrl(baseUrl) {
  try {
    console.error(`Diagnostic: probing ${baseUrl}`);
    const parsed = new URL(baseUrl);
    const host = parsed.hostname || parsed.pathname;
    const port = parsed.port ? Number(parsed.port) : parsed.protocol === "https:" ? 443 : 80;
    console.error(`Resolved host=${host} port=${port} scheme=${parsed.protocol}`);

    try {
      const addrs = await dns.promises.lookup(host, { all: true });
      const uniq = [...new Set(addrs.map((a) => a.address))];
      console.error("Address(es):", uniq.join(", "));
    } catch (e) {
      console.error("dns.lookup failed:", e && e.message ? e.message : e);
    }

    // TCP connect probe
    await new Promise((resolve) => {
      const s = new net.Socket();
      let done = false;
      s.setTimeout(3000);
      s.on("connect", () => {
        done = true;
        console.error("TCP connect: ok");
        s.destroy();
        resolve();
      });
      s.on("error", (err) => {
        if (!done) {
          done = true;
          console.error("TCP connect failed:", err && err.message ? err.message : err);
          resolve();
        }
      });
      s.on("timeout", () => {
        if (!done) {
          done = true;
          console.error("TCP connect timed out");
          s.destroy();
          resolve();
        }
      });
      s.connect(port, host);
    });

    // HTTP health probe
    const healthUrl = baseUrl.replace(/\/+$/, "") + "/health";
    for (let attempt = 1; attempt <= 3; attempt++) {
      try {
        console.error(`HTTP probe attempt ${attempt} -> ${healthUrl}`);
        const controller = new AbortController();
        const id = setTimeout(() => controller.abort(), 5000);
        const r = await fetch(healthUrl, { signal: controller.signal });
        clearTimeout(id);
        console.error("HTTP probe status:", r.status);
        break;
      } catch (e) {
        console.error(`Probe attempt ${attempt} failed:`, e && e.message ? e.message : e);
        await new Promise((r) => setTimeout(r, 500));
      }
    }
  } catch (e) {
    console.error("Diagnostic failure:", e && e.message ? e.message : e);
  }
}

function parseArgs(argv) {
  const out = {};
  const args = argv.slice(2);
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === "--debug") {
      out.debug = true;
      continue;
    }
    if (a.startsWith("--")) {
      const key = a.replace(/^--+/, "").replace(/-/g, "_");
      const val = args[i + 1];
      if (!val || val.startsWith("--")) {
        out[key] = true;
      } else {
        out[key] = val;
        i++;
      }
    }
  }
  return out;
}

async function getSessions(client) {
  try {
    return await client.session.list();
  } catch (e) {
    throw e;
  }
}

async function getLastMessage(client, sessionId) {
  try {
    // preferred signature per SDK: session.messages({ path: { id } })
    const res = await client.session.messages({ path: { id: sessionId } });
    if (Array.isArray(res) && res.length) return res[res.length - 1];
    // some SDK variants may return { data: [...] }
    if (res && Array.isArray(res.data) && res.data.length) return res.data[res.data.length - 1];
  } catch (e) {
    // ignore and fallthrough to alternative
  }

  // fallback: message.list if present
  if (client.message && typeof client.message.list === "function") {
    try {
      const msgs = await client.message.list({ session_id: sessionId });
      if (Array.isArray(msgs) && msgs.length) return msgs[msgs.length - 1];
      if (msgs && Array.isArray(msgs.data) && msgs.data.length) return msgs.data[msgs.data.length - 1];
    } catch (e) {
      // ignore
    }
  }
  return null;
}

function extractModelProviderFromMessage(msg) {
  if (!msg) return [null, null];
  let data = null;
  // SDK message shapes vary; try common locations
  if (msg.info && typeof msg.info === "object") data = { ...msg.info, ...msg };
  else if (msg.metadata && typeof msg.metadata === "object") data = { ...msg.metadata, ...msg };
  else if (msg) data = msg;

  const info = data.info || data.metadata || data.meta || data.extras || data;
  if (info && typeof info === "object") {
    const model_id = info.model_id || info.modelId || info.modelID || info.model || null;
    const provider_id = info.provider_id || info.providerId || info.providerID || info.provider || null;
    return [model_id || null, provider_id || null];
  }
  return [null, null];
}

async function sendMessage({ text, baseUrl, sessionId, modelId, providerId, debug }) {
  const client = createOpencodeClient({ baseUrl });

  try {
    let target = sessionId;
    if (!target) {
      const sessions = await getSessions(client);
      const arr = Array.isArray(sessions) ? sessions : (sessions?.data || []);
      if (!arr || arr.length === 0) {
        console.error("No sessions available on the Opencode instance.");
        return 2;
      }
      const active = arr.find((s) => s.status === "active");
      const selected = active || arr[0];
      target = selected?.id || selected?.session_id || null;
      if (!target) {
        console.error("Could not determine session id from session object:", selected);
        return 3;
      }
    }

    const parts = [{ type: "text", text }];

    if (!modelId || !providerId) {
      const msg = await getLastMessage(client, target);
      const [inferredModel, inferredProvider] = extractModelProviderFromMessage(msg);
      if (inferredModel && inferredProvider) {
        modelId = modelId || inferredModel;
        providerId = providerId || inferredProvider;
        console.error(`Inferred model_id=${modelId} provider_id=${providerId} from last message`);
      } else {
        console.error(
          "Could not infer model_id/provider_id from recent messages. Provide --model-id/--provider-id or set OPENCODE_MODEL_ID/OPENCODE_PROVIDER_ID env vars."
        );
        return 6;
      }
    }

    const modelObj = { providerID: providerId, modelID: modelId };

    // Preferred high-level send
    try {
      if (client.session && typeof client.session.prompt === "function") {
        await client.session.prompt({ path: { id: target }, body: { parts, model: modelObj } });
      } else {
        throw new Error("session.prompt not available");
      }
    } catch (e) {
      console.error(`High-level send failed (${e?.name || "Error"}): ${e?.message || e}; trying low-level POST /session/${target}/prompt`);
      // Fallback to raw POST
      try {
        const url = baseUrl.replace(/\/+$/, "") + `/session/${encodeURIComponent(target)}/prompt`;
        const resp = await fetch(url, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ parts, model: modelObj }),
        });
        // attempt to print response status
        console.error("Low-level POST status:", resp.status);
        if (!resp.ok) {
          const textResp = await resp.text();
          console.error("Low-level POST body:", textResp);
          throw new Error(`POST failed ${resp.status}`);
        }
      } catch (e2) {
        console.error("Low-level POST failed:", e2 && e2.message ? e2.message : e2);
        throw e2;
      }
    }

    console.log(`Message sent to session ${target}`);
    return 0;
  } catch (e) {
    console.error("Error sending message:", e && e.message ? e.message : e);
    if (e && e.stack) console.error(e.stack);
    return 5;
  }
}

async function main() {
  const args = parseArgs(process.argv);
  const text = args.text;
  const baseUrl = args.base_url || process.env.OPENCODE_BASE_URL || "http://127.0.0.1:4096";
  const debug = !!args.debug;
  const modelId = args.model_id || process.env.OPENCODE_MODEL_ID;
  const providerId = args.provider_id || process.env.OPENCODE_PROVIDER_ID;
  const sessionId = args.session_id || null;

  if (!text) {
    console.error("--text is required");
    process.exit(1);
  }

  if (debug) await diagnoseBaseUrl(baseUrl);

  const code = await sendMessage({ text, baseUrl, sessionId, modelId, providerId, debug });
  process.exit(code);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}
